// react-rasta/lib/es6/column/Column.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const theme_1 = require("../theme");
const utils_1 = require("../utils");
require("../utils/bootstrap");
const renderAlignSelf_1 = __importDefault(require("./renderAlignSelf"));
const renderFlex_1 = __importDefault(require("./renderFlex"));
const renderOffset_1 = __importDefault(require("./renderOffset"));
const renderOrder_1 = __importDefault(require("./renderOrder"));
const renderSize_1 = __importDefault(require("./renderSize"));
const Column = theme_1.styled.div `
  position: relative;
  width: 100%;
  min-height: 1px;

  ${(props) => {
    const renderer = {
        alignSelf: (value) => renderAlignSelf_1.default(value),
        flex: (value) => renderFlex_1.default(value),
        offset: (value) => renderOffset_1.default(value),
        order: (value) => renderOrder_1.default(value),
        size: (value) => renderSize_1.default(value),
    };
    const valueMap = {
        alignSelf: props.alignSelf,
        flex: props.flex,
        offset: props.offset,
        order: props.order,
        size: props.size,
    };
    return utils_1.render(valueMap, renderer, props.theme);
}};
`;
exports.default = Column;
