// react-rasta/lib/es5/row/Row.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var theme_1 = require("../theme");
var utils_1 = require("../utils");
require("../utils/bootstrap");
var renderAlignContent_1 = __importDefault(require("./renderAlignContent"));
var renderAlignItems_1 = __importDefault(require("./renderAlignItems"));
var renderDirection_1 = __importDefault(require("./renderDirection"));
var renderGutter_1 = __importDefault(require("./renderGutter"));
var renderJustifyContent_1 = __importDefault(require("./renderJustifyContent"));
var renderWrap_1 = __importDefault(require("./renderWrap"));
var Row = theme_1.styled.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  display: flex;\n\n  ", ";\n"], ["\n  display: flex;\n\n  ",
    ";\n"])), function (props) {
    var renderer = {
        alignContent: function (value) {
            return renderAlignContent_1.default(value);
        },
        alignItems: function (value) {
            return renderAlignItems_1.default(value);
        },
        justifyContent: function (value) {
            return renderJustifyContent_1.default(value);
        },
        direction: function (value) {
            return renderDirection_1.default(value);
        },
        gutter: function (value) {
            return renderGutter_1.default(value, props.theme);
        },
        wrap: function (value) { return renderWrap_1.default(value); },
    };
    var valueMap = {
        alignContent: props.alignContent,
        alignItems: props.alignItems,
        justifyContent: props.justifyContent,
        direction: props.direction,
        gutter: props.noGutter,
        wrap: props.wrap,
    };
    return utils_1.render(valueMap, renderer, props.theme);
});
exports.default = Row;
var templateObject_1;
