// react-rasta/lib/es6/row/Row.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const theme_1 = require("../theme");
const utils_1 = require("../utils");
require("../utils/bootstrap");
const renderAlignContent_1 = __importDefault(require("./renderAlignContent"));
const renderAlignItems_1 = __importDefault(require("./renderAlignItems"));
const renderDirection_1 = __importDefault(require("./renderDirection"));
const renderGutter_1 = __importDefault(require("./renderGutter"));
const renderJustifyContent_1 = __importDefault(require("./renderJustifyContent"));
const renderWrap_1 = __importDefault(require("./renderWrap"));
const Row = theme_1.styled.div `
  display: flex;

  ${(props) => {
    const renderer = {
        alignContent: (value) => renderAlignContent_1.default(value),
        alignItems: (value) => renderAlignItems_1.default(value),
        justifyContent: (value) => renderJustifyContent_1.default(value),
        direction: (value) => renderDirection_1.default(value),
        gutter: (value) => renderGutter_1.default(value, props.theme),
        wrap: (value) => renderWrap_1.default(value),
    };
    const valueMap = {
        alignContent: props.alignContent,
        alignItems: props.alignItems,
        justifyContent: props.justifyContent,
        direction: props.direction,
        gutter: props.noGutter,
        wrap: props.wrap,
    };
    return utils_1.render(valueMap, renderer, props.theme);
}};
`;
exports.default = Row;
